### ISSUE TITLE
Does attention masking actually work?

### ISSUE THREAD

**Issue by Birch-san** [huggingface/diffusers#1890]
I tried passing in an `attention_mask`, for use in a stable-diffusion Unet but it doesn't actually get passed down as deep as `CrossAttention#forward`.

I tried fixing it to pass the param down, but it blows up on tensor size mismatch, because self-attention and cross-attention have different masking requirements.

I made my own implementation of cross-attention masking a few weeks ago (before the refactor) but never upstreamed it. mainly because I didn't understand whether I'd done it correctly (I re-used the [lucidrains implementation](https://github.com/lucidrains/perceiver-pytorch/blob/c8c5f5721520460369a66b8a0e9c5147df4a883e/perceiver_pytorch/perceiver_pytorch.py#L108-L112) that CompVis used):  
https://github.com/huggingface/diffusers/commit/cbb4c02e32ca74b7ac539857c91aa06bd4c9338c  
**EDIT:** rebased implementation to show how it would fit in with the existing attention masking and the refactored attention:  
https://github.com/Birch-san/diffusers/commit/e3a93e9d80a6b4e5122e5b9d02ad4ee60c7d1354

I explicitly named the parameter as a _cross_ attention mask, because a self-attention mask has entirely different requirements.

in terms of wider API design, I wonder whether it should be an attention _map_ (i.e. so you can use it to increase/decrease attention scores for certain token embeds). but for now I'm mostly interested in the mask aspect. because waifu-diffusion makes use of "multiple CLIP embeddings stitched together", so attention masking is useful to avoid attending to padding token embeddings, which would be biased towards conveying high-level semantic of the final CLIP segment only.

@patrickvonplaten

**Comment by patrickvonplaten** [comment]
That's a good point! Also cc @williamberman and @patil-suraj here.

Stable Diffusion never used a attention mask when training, however some other models that make use of `UNet2DCondition` use it (such as UnCLIP). With more and more people fine-tuning stable diffusion we could actually allow `attention_mask` to also be used in Stable Diffusion.

@patil-suraj @williamberman wdyt?

**Comment by Birch-san** [comment]
I see two masking use-cases:

- self-attention mask (if you're training on a batch of images with mixed aspect ratios: you can tell it not to attend to padding pixels)
- cross-attention mask (don't attend to PAD token embeddings in CLIP text condition)

waifu-diffusion already wants to use cross-attention masks for fine-tuning stable-diffusion (in fact will begin a big training run in a few days).

but, what should the API be? `attention_mask` could refer to self-attention _or_ cross-attention. but it'd be catastrophic to pass it to _both_.  
is `cross_attention_kwargs` a useful piece in this puzzle? are we supposed to do something like `cross_attention_kwargs = { 'attention_mask': my_cool_mask }`?

side-note: why is the class named `CrossAttention`? it's super confusing, given that it implements self-attention too. would `MultiHeadAttention` be a better name?  
side-note 2: is there a reason why you didn't use PyTorch's native [`torch.nn.MultiheadAttention`](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html)? you can swap diffusers' `CrossAttention` class for the native implementation [pretty easily](https://github.com/Birch-san/diffusers/compare/fc9e2b76961fab89f59df54475524f9b29d6785b...62388ae7fa04d1c795b909f0d9681850a10b0446).

**Comment by patrickvonplaten** [comment]
Hey @Birch-san,

Thanks for your thoughtful answer here. I see the need to pass customized `attention_masks` . For this IMO the user should set up a customized attention processor class as was merged in: https://github.com/huggingface/diffusers/pull/1639 . Now since we need different attention processors for the self- and cross-attention layers we need to leverage some new functionality that will be added as part of [this](https://github.com/huggingface/diffusers/pull/1884/files#r1065964766) PR (hope to get it merged this week).

As you can see in the comment, it allows to set layers depending on the weight name which should make it to set attention processors only to the self- or cross-attention processors. Does this make sense?

**Comment by Lime-Cakes** [comment]
On the topic of masking when training with padded pixels? Conv doesn't allow mask. Would the padded value be zero and passed into conv, then masked at attention? Would the model still function right when conv might learn the letterbox edges, while attention ignores them?

**Comment by Lime-Cakes** [comment]
> I see two masking use-cases:
> 
>     * self-attention mask (if you're training on a batch of images with mixed aspect ratios: you can tell it not to attend to padding pixels)
> 
>     * cross-attention mask (don't attend to PAD token embeddings in CLIP text condition)
> 
> 
> waifu-diffusion already wants to use cross-attention masks for fine-tuning stable-diffusion (in fact will begin a big training run in a few days).
> 
> but, what should the API be? `attention_mask` could refer to self-attention _or_ cross-attention. but it'd be catastrophic to pass it to _both_. is `cross_attention_kwargs` a useful piece in this puzzle? are we supposed to do something like `cross_attention_kwargs = { 'attention_mask': my_cool_mask }`?
> 
> side-note: why is the class named `CrossAttention`? it's super confusing, given that it implements self-attention too. would `MultiHeadAttention` be a better name? side-note 2: is there a reason why you didn't use PyTorch's native [`torch.nn.MultiheadAttention`](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html)? you can swap diffusers' `CrossAttention` class for the native implementation [pretty easily](https://github.com/Birch-san/diffusers/compare/fc9e2b76961fab89f59df54475524f9b29d6785b...62388ae7fa04d1c795b909f0d9681850a10b0446).

Do you have any update for the result of the training using cross attention mask? Does it work?

**Comment by bonlime** [comment]
want to give +1 to importance of having cross-attention masks. for SD having too much padding tokens affects the image generation (this is probably one of the reason people found longer prompts to work better, to avoid too much attention drawn to the PAD token). I don't see why padding/not padding should change the output image even a tiny bit, so proper support for masks is required

**Comment by patil-suraj** [comment]
It indeed makes sense to pass an attention mask to cross attention.But the stable diffusion model is trained without the masks and to keep the implem 1:1 with the original implem we don't pass attention mask. But it makes sense to allow passing masks optionally.

**Comment by Lime-Cakes** [comment]
Having the option to use it would be great! It should by off be default, to produce same training/inference result as the original implementation.

A lot people are experimenting with fine tuning with more tokens, meaning more padding are used. A few test runs suggest having too much padded tokens fed into unet for training causes lower quality. Personally, I think masking at unet cross attention should be a solution to this issue.

**Comment by bonlime** [comment]
I ended up using such implementation which works nicely. maybe this could make it to the upstream? 
Currently this could only be implemented for default CrossAttention, because xformers do not support custom attention bias to be passed (but I think they are working on it)

```python
class FixedCrossAttnProcessor:
    """Copy-paste from HF diffusers, but with support for passing `encoder_attention_mask` which avoids giving 
    attention to padded tokens"""
    def __call__(
        self,
        attn: CrossAttention,
        hidden_states,
        encoder_hidden_states=None,
        attention_mask=None,
        encoder_attention_mask=None,
    ):
        batch_size, sequence_length, _ = hidden_states.shape
        attention_mask = attn.prepare_attention_mask(attention_mask, sequence_length, batch_size)
        query = attn.to_q(hidden_states)

        if encoder_attention_mask is not None and encoder_hidden_states is not None:
            # B x 77 -> B x 4096 x 77
            attention_mask = encoder_attention_mask.unsqueeze(1).repeat(1, hidden_states.size(1), 1)
            attention_mask = attention_mask.repeat_interleave(attn.heads, dim=0)

        if encoder_hidden_states is None:
            encoder_hidden_states = hidden_states
        elif attn.cross_attention_norm:
            encoder_hidden_states = attn.norm_cross(encoder_hidden_states)

        key = attn.to_k(encoder_hidden_states)
        value = attn.to_v(encoder_hidden_states)

        query = attn.head_to_batch_dim(query)
        key = attn.head_to_batch_dim(key)
        value = attn.head_to_batch_dim(value)

        attention_probs = attn.get_attention_scores(query, key, attention_mask)
        hidden_states = torch.bmm(attention_probs, value)
        hidden_states = attn.batch_to_head_dim(hidden_states)

        # linear proj
        hidden_states = attn.to_out[0](hidden_states)
        # dropout
        hidden_states = attn.to_out[1](hidden_states)

        return hidden_states
```

**Comment by lijuncheng16** [comment]
@Birch-san 
![image](https://github.com/huggingface/diffusers/assets/6412709/f8f74442-7055-4a05-8c78-bb3c616f2c5f)
In our experiment, we find attention mask results in inferior generation compared to unmasked version, and here's our potential examplanation: https://github.com/audiojourney/audiojourney.github.io/blob/main/neurIPS_2023_appendix_v1.3.pdf

**Comment by Ga-Lee** [comment]
> want to give +1 to importance of having cross-attention masks. for SD having too much padding tokens affects the image generation (this is probably one of the reason people found longer prompts to work better, to avoid too much attention drawn to the PAD token). I don't see why padding/not padding should change the output image even a tiny bit, so proper support for masks is required
> 
> upd. it seems that currently the proper masks can be supported by writing a custom `CrossAttnProcessor` and passing `cross_attention_kwargs = { 'attention_mask_': my_cool_mask }`. using the default name `attention_mask` raises errors

Hi, glad to find your answer! I am a green-hand in SD, and I am recently also thinking about the effect of pad tokens during inference. Could please tell me  where I can find the observation or discussion about " _for SD having too much padding tokens affects the image generation_"? Any papers or blogs will be much appreciated! Thank you for guidance!

**Comment by alexblattner** [comment]
I have not seen any examples on how to use this anywhere. would appreciate an example and I'm trying to apply some prompts in some areas of the image in the attention, how does one do that?

**Comment by Birch-san** [comment]
these attention masks are just for hiding padding in the text condition. this interface cannot be used for implementing regional attention.

if you want regional attention, study:  
https://github.com/Birch-san/regional-attn

**Comment by alexblattner** [comment]
@Birch-san thank you very much! I knew sana did something like that, but it makes sense for it to exist in sd1.5 too. Too bad about regional stuff though
